packages/flame/lib/src/effects/controllers/speed_effect_controller.dart

```diff
@@ -0,0 +1,89 @@
+import '../effect.dart';
+import '../measurable_effect.dart';
+import 'duration_effect_controller.dart';
+import 'effect_controller.dart';
+
+/// This controller can force execution of an effect at a predefined speed.
+///
+/// For most of the effect controllers, their duration is set by the user in
+/// advance. This controller is different: it knows the target speed at which
+/// the parent effect wants to proceed, communicates with the effect to
+/// determine its total distance, and uses that to calculate the desired
+/// duration for the [child] effect controller.
+///
+/// Some restrictions apply:
+///   - the [speed] cannot be zero (or negative),
+///   - the [child] controller must be a [DurationEffectController],
+///   - the parent effect must be a [MeasurableEffect].
+class SpeedEffectController extends EffectController {
+  SpeedEffectController(this.child, {required this.speed})
+      : assert(speed > 0, 'Speed must be positive: $speed'),
+        super.empty();
+
+  final DurationEffectController child;
+  final double speed;
+  late MeasurableEffect _parentEffect;
+
+  /// Note that this controller's [started] property is true even if the
+  /// controller is not initialized yet. This is because we want the [Effect]
+  /// to run its `onStart()` callback before we initialize the controller
+  /// (which will happen at the first call to `advance()`).
+  bool _initialized = false;
+
+  @override
+  bool get isRandom => true;
+
+  @override
+  bool get completed => child.completed;
+
+  @override
+  double? get duration => _initialized ? child.duration : double.nan;
+
+  @override
+  double get progress => child.progress;
+
+  @override
+  double advance(double dt) {
+    if (!_initialized) {
+      _initialized = true;
+      final measure = _parentEffect.measure();
```

Comment:
This must become quite expensive?

Reply:
This runs only once at the start of the effect (guarded by `_initialized` flag).